frame/support/test/tests/pallet_compatibility_instance.rs

```diff
@@ -0,0 +1,311 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use sp_runtime::traits::Block as _;
+
+mod pallet_old {
+	use frame_support::{
+		decl_storage, decl_error, decl_event, decl_module, weights::Weight, traits::Get, Parameter
+	};
+	use frame_system::ensure_root;
+
+	pub trait Trait<I: Instance = DefaultInstance>: frame_system::Trait {
+		type SomeConst: Get<Self::Balance>;
+		type Balance: Parameter + codec::HasCompact + From<u32> + Into<Weight> + Default;
+		type Event: From<Event<Self, I>> + Into<<Self as frame_system::Trait>::Event>;
+	}
+
+	decl_storage! {
+		trait Store for Module<T: Trait<I>, I: Instance = DefaultInstance> as Example {
+			/// Some documentation
+			Dummy get(fn dummy) config(): Option<T::Balance>;
+			Bar get(fn bar) config(): map hasher(blake2_128_concat) T::AccountId => T::Balance;
+			Foo get(fn foo) config(): T::Balance = 3.into();
+			Double get(fn double): double_map hasher(blake2_128_concat) u32, hasher(twox_64_concat) u64 => u16;
+		}
+	}
+
+	decl_event!(
+		pub enum Event<T, I = DefaultInstance> where Balance = <T as Trait<I>>::Balance {
+			/// Dummy event, just here so there's a generic type that's used.
+			Dummy(Balance),
+		}
+	);
+
+	decl_module! {
+		pub struct Module<T: Trait<I>, I: Instance = DefaultInstance> for enum Call where origin: T::Origin {
+			type Error = Error<T, I>;
+			fn deposit_event() = default;
+			const SomeConst: T::Balance = T::SomeConst::get();
+
+			#[weight = <T::Balance as Into<Weight>>::into(new_value.clone())]
+			fn set_dummy(origin, #[compact] new_value: T::Balance) {
+				ensure_root(origin)?;
+
+				<Dummy<T, I>>::put(&new_value);
+				Self::deposit_event(RawEvent::Dummy(new_value));
+			}
+
+			fn on_initialize(_n: T::BlockNumber) -> Weight {
+				<Dummy<T, I>>::put(T::Balance::from(10));
+				10
+			}
+
+			fn on_finalize(_n: T::BlockNumber) {
+				<Dummy<T, I>>::put(T::Balance::from(11));
+			}
+		}
+	}
+
+	decl_error! {
+		pub enum Error for Module<T: Trait<I>, I: Instance> {
+			/// Some wrong behavior
+			Wrong,
+		}
+	}
+}
+
+#[frame_support::pallet(Example)]
```

Comment:
Do we may directly want to switch to `PalletInfo` to fetch this name?

Reply:
Indeed we can do that, then Instance doesn't need to contain any information as the `PalletInfo::name` is already unique.

Thus I would remove all generated instance, change `= DefaultInstance` (which avoid breaking when converting non-instantiable to instantiable) to `= ()`.
So the generic I would just be parameter to be able to implement the pallet multiple time, but no information would be needed from it anymore.

Another note: some pallets such as collective will need to migrate or be renamed in construct_runtime to `Instance1Collective`, maybe we can a method to PalletInfo `fn pallet_prefix` which is by default pallet name or configurable in construct_runtime.

Reply:
> Another note: some pallets such as collective will need to migrate or be renamed in construct_runtime to Instance1Collective, maybe we can a method to PalletInfo fn pallet_prefix which is by default pallet name or configurable in construct_runtime.

I don't understand this, for what do we need `pallet_prefix`?

And with the migration you are right, but that will be do able. I would assume that after merging this, we will move over the pallets one by one to the new proc-macro.

Reply:
If user want to call his pallet Council while still having the storage using the prefix `instance1Collective` then maybe in construct_runtime we can do:
```
#[pallet_storage_prefix = "instance1Collective"]
Council: pallet_collective::<Instance1>::{Module, Call, Storage, Origin<T>, Event<T>, Config<T>},
```
which would implement `PalletInfo::name` with "Council" and `PalletInfo::storage_prefix` with "instance1Collective".
But if we migrate everything this feature is useless. (and collective should be easy to migrate anyway)

Reply:
This is done, now storages make use of PalletInfo for finding the module_prefix to use.

Reply:
Thus macro no longe require to bound Instance trait, but only `'static`. Instance trait is no longer implemented for the generated instance placeholder (`struct Instance1;`, `struct Instance2;`, ..).
The DefaultInstance is just `()`.